### PR TITLE
feat: Implement get and update draft article for non existing page - EXO-69697_69639

### PR DESCRIPTION
**Pull request by sofyenne**

### CODE REVIEW

File: services/src/main/java/org/exoplatform/news/service/NewsService.java [hunk 1]
```diff
@@ -59,6 +59,8 @@ public interface NewsService {
    * @throws Exception
    */
   News updateNews(News news, String updater, Boolean post, boolean publish) throws Exception;
+
+  News updateNews(News news, String updater, Boolean post, boolean publish, String newsType) throws Exception;
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/service/NewsService.java]:
```suggestion
  News updateNews(News news, String updater, Boolean post, boolean publish, String newsObjectType) throws Exception;
```
add java doc here

File: services/src/main/java/org/exoplatform/news/rest/NewsRestResourcesV1.java [hunk 2]
```diff
@@ -689,10 +691,11 @@ public Response getNewsIllustration(@Context Request rsRequest,
                                       @Context HttpServletRequest request,
                                       @Parameter(description = "News id", required = true) @PathParam("id") String id,
                                       @Parameter(description = "last modified date") @QueryParam("v") long lastModified,
+                                      @Parameter(description = "news type to be fetched", required = false) @QueryParam("type") String newsType,
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsRestResourcesV1.java]:
```suggestion
                                      @Parameter(description = "News object type to be fetched", required = false) @QueryParam("type") String newsObjectType,
```

File: services/src/main/java/org/exoplatform/news/rest/NewsRestResourcesV1.java [hunk 3]
```diff
@@ -360,14 +361,15 @@ public Response undoDeleteNews(
   public Response getNewsById(@Context HttpServletRequest request,
                               @Parameter(description = "News id", required = true) @PathParam("id") String id,
                               @Parameter(description = "fields", required = true) @QueryParam("fields") String fields,
+                              @Parameter(description = "News type to be fetched", required = false) @QueryParam("type") String newsType,
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsRestResourcesV1.java]:
```suggestion
                              @Parameter(description = "News object type to be fetched", required = false) @QueryParam("type") String newsObjectType,
```

File: services/src/main/java/org/exoplatform/news/rest/NewsRestResourcesV1.java [hunk 4]
```diff
@@ -214,14 +214,15 @@ public Response canCreateNews(@Context HttpServletRequest request,
       @ApiResponse(responseCode = "500", description = "Internal server error") })
   public Response updateNews(@Parameter(description = "News id", required = true) @PathParam("id") String id,
                              @Parameter(description = "Post news", required = false) @QueryParam("post") boolean post,
+                             @Parameter(description = "News type to be updated", required = false) @QueryParam("type") String newsType,
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsRestResourcesV1.java]:
```suggestion
                             @Parameter(description = "News object type to be updated", required = false) @QueryParam("type") String newsObjectType,
```

File: services/src/main/java/org/exoplatform/news/rest/NewsRestResourcesV1.java [hunk 5]
```diff
@@ -214,14 +214,15 @@ public Response canCreateNews(@Context HttpServletRequest request,
...
                              @RequestBody(description = "News object to be updated", required = true) News updatedNews) {
 
     if (updatedNews == null) {
       return Response.status(Response.Status.BAD_REQUEST).build();
     }
     org.exoplatform.services.security.Identity currentIdentity = ConversationState.getCurrent().getIdentity();
     try {
-      News news = newsService.getNewsById(id, currentIdentity, false);
+      News news = newsService.getNewsById(id, currentIdentity, false, newsType);
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsRestResourcesV1.java]:
```suggestion
      News news = newsService.getNewsById(id, currentIdentity, false, newsObjectType);
```

File: services/src/main/java/org/exoplatform/news/rest/NewsRestResourcesV1.java [hunk 6]
```diff
@@ -237,7 +238,7 @@ public Response updateNews(@Parameter(description = "News id", required = true)
       news.setAudience(updatedNews.getAudience());
 
 
-      news = newsService.updateNews(news, currentIdentity.getUserId(), post, updatedNews.isPublished());
+      news = newsService.updateNews(news, currentIdentity.getUserId(), post, updatedNews.isPublished(), newsType);
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsRestResourcesV1.java]:
```suggestion
      news = newsService.updateNews(news, currentIdentity.getUserId(), post, updatedNews.isPublished(), newsObjectType);
```

File: services/src/main/java/org/exoplatform/news/rest/NewsRestResourcesV1.java [hunk 7]
```diff
@@ -360,14 +361,15 @@ public Response undoDeleteNews(
...
+                              @Parameter(description = "News type to be fetched", required = false) @QueryParam("type") String newsType,
                               @Parameter(description = "Is edit mode") @Schema(defaultValue = "false") @QueryParam("editMode") boolean editMode) {
     String authenticatedUser = request.getRemoteUser();
     try {
       if (StringUtils.isBlank(id)) {
         return Response.status(Response.Status.BAD_REQUEST).build();
       }
       org.exoplatform.services.security.Identity currentIdentity = ConversationState.getCurrent().getIdentity();
-      News news = newsService.getNewsById(id, currentIdentity, editMode);
+      News news = newsService.getNewsById(id, currentIdentity, editMode, newsType);
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsRestResourcesV1.java]:
```suggestion
      News news = newsService.getNewsById(id, currentIdentity, editMode, newsObjectType);
```

File: services/src/main/java/org/exoplatform/news/rest/NewsRestResourcesV1.java [hunk 8]
```diff
@@ -689,10 +691,11 @@ public Response getNewsIllustration(@Context Request rsRequest,
                                       @Context HttpServletRequest request,
                                       @Parameter(description = "News id", required = true) @PathParam("id") String id,
                                       @Parameter(description = "last modified date") @QueryParam("v") long lastModified,
+                                      @Parameter(description = "news type to be fetched", required = false) @QueryParam("type") String newsType,
                                       @Parameter(description = "resized image size") @QueryParam("size") String size) {
     try {
       org.exoplatform.services.security.Identity currentIdentity = ConversationState.getCurrent().getIdentity();
-      News news = newsService.getNewsById(id, currentIdentity, false);
+      News news = newsService.getNewsById(id, currentIdentity, false, newsType);
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsRestResourcesV1.java]:
```suggestion
      News news = newsService.getNewsById(id, currentIdentity, false, newsObjectType);
```

File: services/src/main/java/org/exoplatform/news/service/NewsService.java [hunk 9]
```diff
@@ -249,4 +251,7 @@ public interface NewsService {
    * @return boolean : true if user can archive the news
    */
   boolean canArchiveNews(org.exoplatform.services.security.Identity currentIdentity, String newsAuthor);
+
+  News getNewsById(String newsId, org.exoplatform.services.security.Identity currentIdentity, boolean editMode, String newsType) throws IllegalAccessException;
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/service/NewsService.java]:
```suggestion
  News getNewsById(String newsId, org.exoplatform.services.security.Identity currentIdentity, boolean editMode, String newsObjectType) throws IllegalAccessException;
```
add java doc here

File: services/src/main/java/org/exoplatform/news/service/impl/NewsServiceImpl.java [hunk 10]
```diff
@@ -155,8 +155,12 @@ public boolean canCreateNews(Space space, org.exoplatform.services.security.Iden
    */
   @Override
   public News updateNews(News news, String updater, Boolean post, boolean publish) throws Exception {
+    return updateNews(news, updater, post, publish, null);
+  }
+
+  public News updateNews(News news, String updater, Boolean post, boolean publish, String newsType) throws Exception {
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/service/impl/NewsServiceImpl.java:153]:
```suggestion
  public News updateNews(News news, String updater, Boolean post, boolean publish, String newsObjectType) throws Exception {
```

File: services/src/main/java/org/exoplatform/news/service/impl/NewsServiceImpl.java [hunk 11]
```diff
@@ -230,6 +234,11 @@ public void deleteNews(String newsId, org.exoplatform.services.security.Identity
    */
   @Override
   public News getNewsById(String newsId, org.exoplatform.services.security.Identity currentIdentity, boolean editMode) throws IllegalAccessException {
+   return getNewsById(newsId, currentIdentity, editMode, null);
+  }
+
+  @Override
+  public News getNewsById(String newsId, org.exoplatform.services.security.Identity currentIdentity, boolean editMode, String newsType) throws IllegalAccessException {
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/service/impl/NewsServiceImpl.java:249]:
```suggestion
  public News getNewsById(String newsId, org.exoplatform.services.security.Identity currentIdentity, boolean editMode, String newsObjectType) throws IllegalAccessException {
```

File: services/src/main/java/org/exoplatform/news/service/impl/NewsServiceImplV2.java [hunk 12]
```diff
@@ -51,6 +64,8 @@
 import org.exoplatform.wiki.model.Page;
 import org.exoplatform.wiki.service.NoteService;
 
+import static org.exoplatform.news.utils.NewsUtils.NewsObjectType.*;
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/service/impl/NewsServiceImplV2.java]: to be replaced with explicit imports

File: services/src/main/java/org/exoplatform/news/service/impl/NewsServiceImplV2.java [hunk 13]
```diff
@@ -133,6 +162,60 @@ public boolean canCreateNews(Space space, org.exoplatform.services.security.Iden
   public News updateNews(News news, String updater, Boolean post, boolean publish) throws Exception {
     return null;
   }
+  public News updateNews(News news, String updater, Boolean post, boolean publish, String newsObjectType) throws Exception {
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/service/impl/NewsServiceImplV2.java:177]:
```suggestion

  @Override
  public News updateNews(News news, String updater, Boolean post, boolean publish, String newsObjectType) throws Exception {
```

File: services/src/main/java/org/exoplatform/news/service/impl/NewsServiceImplV2.java [hunk 14]
```diff
@@ -238,6 +321,30 @@ public boolean canArchiveNews(Identity currentIdentity, String newsAuthor) {
...
 
+  @Override
+  public News getNewsById(String newsId, Identity currentIdentity, boolean editMode, String newsType) throws IllegalAccessException {
+    News news = null;
+    try {
+      if (newsType == null) {
+        throw new IllegalArgumentException("required argument news type could not be null");
+      }
+      if (NewsUtils.NewsObjectType.DRAFT.name().toLowerCase().equals(newsType)){
+        news = getNewsDraft(newsId, currentIdentity);
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/service/impl/NewsServiceImplV2.java]:
```suggestion
        news = getDraftNews(newsId, currentIdentity);
```

File: services/src/main/java/org/exoplatform/news/service/impl/NewsServiceImplV2.java [hunk 15]
```diff
@@ -252,14 +359,17 @@ public News createDraftArticleForNewPage(News news, String pageOwnerId) throws E
...
+        Map<String, String> properties = new HashMap<>();
         // save illustration
-        Long illustrationId = -1L;
+        Long illustrationId = null;
         if (news.getUploadId() != null && !news.getUploadId().isBlank()) {
           illustrationId = saveNewsIllustration(news.getUploadId(), null);
+          properties.put(NEWS_ILLUSTRATION_ID, String.valueOf(illustrationId));
+          properties.put(NEWS_UPLOAD_ID, news.getUploadId());
+        }
+        if (news.getSummary() != null && !news.getSummary().isBlank()) {
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/service/impl/NewsServiceImplV2.java]: you can use StringUtils.isNotEmpty instead in all same cases

File: services/src/main/java/org/exoplatform/news/service/impl/NewsServiceImplV2.java [hunk 16]
```diff
@@ -319,7 +429,161 @@ private Long saveNewsIllustration(String uploadId, Long oldFileId) {
     }
   }
 
+  public News getNewsDraft(String newsId, Identity currentIdentity) throws WikiException, IllegalAccessException {
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/service/impl/NewsServiceImplV2.java]:
```suggestion
  private News getNewsDraft(String newsId, Identity currentIdentity) throws WikiException, IllegalAccessException {
```

File: services/src/main/java/org/exoplatform/news/service/impl/NewsServiceImplV2.java [hunk 17]
```diff
@@ -319,7 +429,161 @@ private Long saveNewsIllustration(String uploadId, Long oldFileId) {
...
+        StringBuilder newsUrl = new StringBuilder("");
+        newsUrl.append("/").append(PortalContainer.getCurrentPortalContainerName()).append("/").append(CommonsUtils.getCurrentPortalOwner()).append("/news/detail?newsId=").append(draftArticle.getId()).append("&type=draft");
+        draftArticle.setUrl(newsUrl.toString());
+      }
+      return draftArticle;
+    }
+    return null;
+  }
+
+  public News updateNews(News news , String updater, String newsObjectType) throws WikiException, FileStorageException, IllegalAccessException {
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/service/impl/NewsServiceImplV2.java]:
```suggestion
  private News updateNews(News news , String updater, String newsObjectType) throws WikiException, FileStorageException, IllegalAccessException {
```

File: services/src/main/java/org/exoplatform/news/service/impl/NewsServiceImplV2.java [hunk 18]
```diff
@@ -319,7 +429,161 @@ private Long saveNewsIllustration(String uploadId, Long oldFileId) {
     }
   }
 
+  public News getNewsDraft(String newsId, Identity currentIdentity) throws WikiException, IllegalAccessException {
+    return buildNewsDraftFromDraftPage(newsId, currentIdentity.getUserId());
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/service/impl/NewsServiceImplV2.java]: why creating another method ?

File: services/src/main/java/org/exoplatform/news/service/impl/NewsServiceImplV2.java [hunk 19]
```diff
@@ -319,7 +429,161 @@ private Long saveNewsIllustration(String uploadId, Long oldFileId) {
...
+        }
+        existingMetadata.setProperties(properties);
+        String identityId = identityManager.getOrCreateUserIdentity(updater).getId();
+        metadataService.updateMetadataItem(existingMetadata, Long.parseLong(identityId));
+      }
+    }
+    return news;
+  }
+
+  private boolean canEditNews(News news, String authenticatedUser) {
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/service/impl/NewsServiceImplV2.java:617]: this method is updated in NewsServiceImpl to be updated here

File: services/src/test/java/org/exoplatform/news/service/impl/NewsServiceImplV2Test.java [hunk 20]
```diff
@@ -121,4 +159,113 @@ public void createNews() throws Exception {
     assertEquals(draftPage.getContent(), savedDraftArticle.getBody());
     assertEquals(draftPage.getAuthor(), savedDraftArticle.getAuthor());
   }
+
+  @Test
+  public void getNewsById() throws Exception {
```
Comment (azayati) [inline services/src/test/java/org/exoplatform/news/service/impl/NewsServiceImplV2Test.java]:
```suggestion
  public void testGetDraftNewsById() throws Exception {
```

File: services/src/test/java/org/exoplatform/news/service/impl/NewsServiceImplV2Test.java [hunk 21]
```diff
@@ -121,4 +159,113 @@ public void createNews() throws Exception {
...
+    assertEquals(draftPage.getAuthor(), news.getAuthor());
+    assertEquals(draftPage.getContent(), news.getBody());
+    assertEquals("draft", news.getPublicationState());
+    assertEquals(space.getDisplayName(), news.getSpaceDisplayName());
+    assertEquals(space.getAvatarUrl(), news.getSpaceAvatarUrl());
+    assertEquals("/portal/dw/news/detail?newsId=1&type=draft", news.getUrl());
+  }
+
+  @Test
+  public void updateNews() throws Exception {
```
Comment (azayati) [inline services/src/test/java/org/exoplatform/news/service/impl/NewsServiceImplV2Test.java]:
```suggestion
  public void testUpdateDraftNews() throws Exception {
```

**Review by azayati — CHANGES_REQUESTED** [review]

File: services/src/main/java/org/exoplatform/news/service/NewsService.java (hunk 1 above)
Comment (sofyenne) [reply services/src/main/java/org/exoplatform/news/service/NewsService.java]: done!

File: services/src/main/java/org/exoplatform/news/service/impl/NewsServiceImplV2.java (hunk 18 above)
Comment (sofyenne) [reply services/src/main/java/org/exoplatform/news/service/impl/NewsServiceImplV2.java]: done !

File: services/src/main/java/org/exoplatform/news/service/impl/NewsServiceImplV2.java (hunk 19 above)
Comment (sofyenne) [reply services/src/main/java/org/exoplatform/news/service/impl/NewsServiceImplV2.java:617]: method updated !

**Review by azayati — APPROVED** [review]
LGTM

**Pull request merged**